### PR TITLE
Adding joint pos control option to continuous grasper

### PR DESCRIPTION
**Pull request by girifb**
# Description

We would like to control the arm via joint positions while performing the continuous grasp trajectory.

Fixes # (issue)

## Type of change

Please check the options that are relevant.

- [ ] Bug fix (non-breaking change that fixes an issue)
- [ ] Proposes a change (non-breaking change that isn't necessarily a bug)
- [ ] Refactor
- [ ] New feature (non-breaking change that adds a new functionality)
- [ ] Breaking change (fix or feature that would break some existing functionality downstream)
- [ ] This is a unit test
- [ ] Documentation only change
- [ ] Datasets Release
- [ ] Models Release

## Type of requested review

- [ ] I want a thorough review of the implementation.
- [ ] I want a high level review. 
- [ ] I want a deep design review.

## Before and After

If this PR introduces a change or fixes a bug, please add before and after screenshots (if this is causes a UX change) or before and after examples(this could be plain text, videos etc ) to demonstrate the change.

If this PR introduces a new feature, please also add examples or screenshots demonstrating the feature.

# Testing

Please describe the tests that you ran to verify your changes and how we can reproduce.

# Checklist:

- [ ] I have performed manual end-to-end testing of the feature in my environment.
- [ ] I have added Docstrings and comments to the code.
- [ ] I have made changes to existing documentation where needed.
- [ ] I have added tests that show that the PR is functional.
- [ ] New and existing unit tests pass locally with my changes.
- [ ] I have added relevant collaborators to review the PR before merge.
- [ ] [Polymetis only] I ran on hardware (1) all scripts in `tests/scripts`, (2) asv benchmarks.

### CODE REVIEW

**Review by exhaustin — COMMENTED** [review]
Since we're updating the desired pose at a high frequency, I'm worried about the performance of the inverse kinematics function (since IK is basically an optimization problem that's iteratively solved).

A better solution would be to use `toco.planning.generate_cartesian_target_joint_min_jerk` instead of `toco.planning.generate_cartesian_space_min_jerk` to solve for IK offline, and also faster since it solves IK locally between adjacent waypoints.

I actually have this implemented a while ago before Cartesian impedance was fast enough: https://github.com/facebookresearch/fairo/blob/96c236875631599ca9f6e835b5201a9047a7d5a7/polymetis/polymetis/python/polymetis/utils/continuous_grasper.py

**Comment by girifb** [comment]
@exhaustin 

- I did try to use `toco.planning.generate_cartesian_target_joint_min_jerk`
- However it returns waypoints in cartesian domain (the tensors are 3x1 instead of 7x1) and i cannot use that with `JointImpedenceControl`
- That is when i chose to use IK API directly

If i can somehow make `JointImpedenceControl` work with `generate_cartesian_target_joint_min_jerk` way points that would be great

**Comment by girifb** [comment]
@exhaustin 

- I was wrong about 3 vs 7. 
- I did try `generate_cartesian_target_joint_min_jerk` and i get core dump

```
python3: /home/robopen02/.conda/envs/polymetis-local/include/eigen3/Eigen/src/Core/DenseCoeffsBase.h:162: Eigen::DenseCoeffsBase<Derived, 0>::CoeffReturnType Eigen::DenseCoeffsBase<Derived, 0>::operator[](Eigen::Index) const [with Derived = Eigen::Matrix<double, -1, 1>; Eigen::DenseCoeffsBase<Derived, 0>::CoeffReturnType = const double&; Eigen::Index = long int]: Assertion `index >= 0 && index < size()' failed.
Aborted (core dumped)
```

**Comment by girifb** [comment]
Thanks @exhaustin @1heart 

This works now and is ready for review

File: polymetis/polymetis/python/polymetis/utils/continuous_grasper.py [hunk 1]
```diff
@@ -107,23 +139,41 @@ def move_to(
...
+                        }
+                    )
+                    if gather_arm_state_func:
+                        observed_state = gather_arm_state_func(
+                            self.arm.get_robot_state()
+                        )
+                        observed_state["joint_pos_desired"] = joint_pos_desired
+                        robot_states.append(observed_state)
+                else:
+                    assert "invalid controller type"
```
Comment (1heart) [inline polymetis/polymetis/python/polymetis/utils/continuous_grasper.py:176]:
```suggestion
                    raise Exception(f"Invalid controller type {self._controller_type}")
```

File: polymetis/polymetis/python/polymetis/utils/continuous_grasper.py [hunk 2]
```diff
@@ -68,14 +72,30 @@ def reset_policy(self):
...
+                joint_pos_current=joint_pos_current,
+                Kp=(self._gain_multiplier * torch.Tensor(self.arm.metadata.default_Kq)),
+                Kd=(
+                    self._gain_multiplier * torch.Tensor(self.arm.metadata.default_Kqd)
+                ),
+                robot_model=self.arm.robot_model,
+            )
+            self.arm.send_torch_policy(policy, blocking=False)
+        else:
+            assert "invalid controller type"
```
Comment (1heart) [inline polymetis/polymetis/python/polymetis/utils/continuous_grasper.py:98]:
```suggestion
            raise Exception(f"Invalid controller type {self._controller_type}")
```

**Review by 1heart — APPROVED** [review]

**Pull request merged**